### PR TITLE
Add additional documentation and builder APIs to `SortOptions`

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Related to https://github.com/apache/datafusion/issues/12446

# Rationale for this change
 
While working on https://github.com/apache/datafusion/pull/12562 I found I kept having to do mental gymnastics to remember the default value of `SortOptions`

I also found creating them was a bit confuisng as if I want to make `ASC` the default I have to do
`descending: false` and the double negative was making my head hurt.

# What changes are included in this PR?
1. Add additional documentation
2. Add builder style APIs that permits configuring an `ASC` sort with `options.asc()` rather than `descending:false`
3. Update some uses to the new builder style to demonstrate its use

# Are there any user-facing changes?
New API and docs, no changes to existing APIs

### CODE REVIEW

File: arrow-schema/src/lib.rs [hunk 1]
```diff
@@ -34,6 +36,35 @@ use std::ops;
...
+///
+/// // Configure using builder APIs
+/// let options = SortOptions::default()
+///  .desc()
+///  .with_nulls_first(true);
+/// assert_eq!(options.to_string(), "DESC NULLS FIRST");
+/// ```
+///
+/// # Example operations
+/// It is also possible to negate the sort options using the `!` operator.
```
Comment (alamb) [inline arrow-schema/src/lib.rs:62]: I found the use of the `!` in DataFusion a bit surprising at first

File: arrow-schema/src/lib.rs [hunk 2]
```diff
@@ -42,6 +73,56 @@ pub struct SortOptions {
...
 
+impl Display for SortOptions {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        if self.descending {
+            write!(f, "DESC")?;
+        } else {
+            write!(f, "ASC")?;
+        }
+        if self.nulls_first {
+            // purposely don't display default NULLs value
```
Comment (etseidl) [inline arrow-schema/src/lib.rs:84]: The default for `nulls_first` doesn't seem to be documented, so wouldn't it be clearer to print the value here?

**Review by etseidl — APPROVED** [review]
Thanks @alamb, looks like a nice addition to me. Just one little nit.

File: arrow-schema/src/lib.rs [hunk 3]
```diff
@@ -34,6 +36,35 @@ use std::ops;
...
+/// };
+/// // Default is ASC NULLs First
+/// assert_eq!(options, SortOptions::default());
+/// assert_eq!(options.to_string(), "ASC");
+///
+/// // Configure using builder APIs
+/// let options = SortOptions::default()
+///  .desc()
+///  .with_nulls_first(true);
+/// assert_eq!(options.to_string(), "DESC NULLS FIRST");
```
Comment (etseidl) [inline arrow-schema/src/lib.rs:58]:
```suggestion
/// assert_eq!(options.to_string(), "DESC");
```
If `fmt` continues to print nothing for the default.

File: arrow-schema/src/lib.rs (hunk 2 above)
Comment (alamb) [reply arrow-schema/src/lib.rs:84]: Yes, I think you are right -- it is more consistent even if more verbose

File: arrow-schema/src/lib.rs [hunk 4]
```diff
@@ -42,6 +73,56 @@ pub struct SortOptions {
...
+    }
+
+    /// Set this sort options to sort descending if argument is true
+    pub fn with_descending(mut self, descending: bool) -> Self {
+        self.descending = descending;
+        self
+    }
+
+    /// Set this sort options to sort nulls first if argument is true
+    pub fn with_nulls_first(mut self, nulls_first: bool) -> Self {
```
Comment (berkaysynnada) [inline arrow-schema/src/lib.rs:120]: We can keep this API for sure, but to be consistent with what you propose at https://github.com/apache/datafusion/pull/12595 (`nulls_first()` / `nulls_last()` API for `PhysicalSortExpr`), should we add `nulls_first()` and `nulls_last()` here, too?

Comment (alamb) [reply arrow-schema/src/lib.rs:120]: it is a good idea -- I will add it

**Comment by alamb** [comment]
Close/reopen to trigger CI

**Comment by alamb** [comment]
integration CI is failing due to https://github.com/apache/arrow-rs/issues/6448

**Comment by alamb** [comment]
Thanks again @etseidl and @berkaysynnada  for the review

**Pull request merged**